---
title: "dna_analysis"
output:
  html_document:
    toc: true
    number_sections: true
  html_notebook: default
---

First we're going to set up R to to our target directory and clear the workspace.
```{r setup}
knitr::opts_chunk$set(echo = TRUE,
                      warning = FALSE, 
                      message = FALSE,
                      engine = "R", # Chunks will always have R code, unless noted
                      error = TRUE,
                      fig.path="Figures/",  # Set the figure options
                      fig.align = "center",
                      dev='png')

```

Clear RStudio
```{r clear}
  # cat(rep("\n",50)) # Clear Console
  rm(list=ls(all=TRUE)) # clear workspace
  graphics.off() # closes all graphics
```

Install functions
```{r }
source("R/functions.R")
```
Install packages
```{r}
packages(tidyverse)
packages(broom)
packages(devtools) # List additional needed packages here on new lines with "packages(NAME)"
packages(survival) # dependency for microbiome package
packages(Formula) # dependency for microbiome package
packages(ggplot2) # dependency for microbiome package
packages(acepack) # dependency for microbiome package
packages(base64enc) # dependency for microbiome package
packages(gridExtra) # dependency for microbiome package
packages(htmltools)
packages(stringi)
packages(htmlTable) # dependency for microbiome package
packages(data.table) # dependency for microbiome package
packages(permute) # dependency for microbiome package
packages(ggthemes)
packages(GGally)
packages(survival)
source("https://bioconductor.org/biocLite.R")
# biocLite("WGCNA")
# biocLite("preprocessCore")
# install_github("jtclaypool/microbiome")
#missing dependencies may need to be installed from Bioconductor
#install missing packages by (for preprocessCore):
#source("https://bioconductor.org/biocLite.R")
#biocLite("preprocessCore")
packages(microbiome)
```

```{r get data, message = FALSE, warning=FALSE}
#metadata

meta <- read_csv("data/meta.csv") %>% filter(!is.na(`Sample Name`))
#filter for samples of interest (anode here)
meta <- meta %>% filter(str_sub(`Sample Label`,-1,-1)=="A")
# meta <- meta %>% filter(str_sub(`Sample Label`,-1,-1)=="A" | str_sub(`Sample Label`,-1,-1)==1|str_sub(`Sample Label`,-1,-1)==2 )
#paste on suffixes used by sequencing company
sample_labels_of_interest <- meta %>% select(`Sample Label`) %>% unlist %>% paste(".515yF.926R",sep = "")
#filter dna seq data for samples of interest and read in
dna_seq_of_interest <- read_tsv("data/tsv_otu_table.txt",skip=1) %>% 
  select(`#OTU ID`,sample_labels_of_interest,ConsensusLineage)

# have to add in first line "# Constructed from biom file"
cat("# Constructed from biom file","\n",file="output/dna_seq_of_interest.txt")
write.table(dna_seq_of_interest,file="output/dna_seq_of_interest.txt",append = TRUE,row.names=FALSE,quote = FALSE,sep = "\t")

biom <- read.biom("output/dna_seq_of_interest.txt")
rm(dna_seq_of_interest)
```

Fix Metadata
```{r fix metadata}
# This section fixes the metadata to match between the metadata labels and the R biome labels
  # Note that R puts an X in front of any column header that starts with a number, and dashes and other special characters are replaced by periods
labels <- meta %>% select(`Sample Label`) %>% unlist %>% paste(".515yF.926R",sep = "") 
meta$`Sample Label`<- ifelse(!is.na(as.numeric(str_sub(labels,1,1))), paste("X",labels,sep = ""), paste("",labels,sep = "")) %>% gsub("-",".", .)
```

Make metadata resemble Exp23 RSM data set of 15 samples
```{r}
meta <- meta %>% mutate(mfc=as.numeric(str_extract(`Sample Name`,"\\d+")))
meta <- meta %>% select(`Sample Label`,mfc) %>% left_join(.,read_csv("rsm_hist.csv")[,1:4])
meta <- meta %>% dplyr::rename(Label = `Sample Label`)
meta <- as.data.frame(meta) #microbiome package doesn't always play nice with tibbles
```

#Sequencing depth and rarefaction curves
```{r}
# https://rdrr.io/rforge/vegan/man/rarefy.html
packages(vegan)
S <- specnumber(biom$RA.Otus) # observed number of species

# Modify data so that rows are samples, columns are reads per OTU
rare.data <- biom$biom_tab %>% select(starts_with("X")) # leave off taxonomy on biom_tab
t.rare.data <- transpose(rare.data) # transpose https://stackoverflow.com/questions/6778908/r-transposing-a-data-frame
# get row and colnames in order
  colnames(t.rare.data) <- rownames(rare.data)
  rownames(t.rare.data) <- colnames(rare.data)

(raremax <- min(rowSums(t.rare.data)))
Srare <- rarefy(t.rare.data, raremax)
plot(S, Srare, xlab = "Observed No. of Species", ylab = "Rarefied No. of Species")
abline(0, 1)
rarecurve(t.rare.data, step = 20, sample = raremax, col = "blue", cex = 0.6)
```
```{r get total reads per sample sequenced}
# make a long form table we can aggregate
ix <- which(colnames(biom$biom_tab) %in% c("Kingdom","Phylum","Class","Order","Family","Genus","Species")) # remove taxonomy info
data <- biom$biom_tab[,-ix] 
data <- t(data)
data <- melt(data)
colnames(data) <- c("Label","OTU","Reads")
# aggregate reads by label
agg_data_label <- aggregate(data$Reads,
    by = list(Label = data$Label),
    FUN = function(x) c(sum = sum(x)))
ggplot(agg_data_label, aes(Label,x))+geom_bar(stat="identity")+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+ylab("Reads")
```




#Sample vs OTU
```{r Simple Relative Abundance}
# Here we can make our generic barplot of relative abundance. We can also then transform it into something fancier!
ra_plot=barplot_RA(biom$RA.Otus,tax = biom$taxon,meta = meta,category = "temp",top = 5)

# plot is stored in list variable RA_plot
# RA_plot is a ggplot2 object and can be manipulated according to make publication ready graph
# for example adding an x-axis label and changing the legend title

ggplot(data = ra_plot$top_long,aes(x=factor(label,levels = c("-1","0","1")),y=value,fill=variable))+ # this reorders x     axis values
  geom_bar(stat="identity")+
  scale_x_discrete("Temperature")+ # Labels categorical labels on x axis
  ylab("Relative Abundance (%)")+ 
  theme(legend.title=element_text(),legend.position="right",plot.title = element_text(hjust=0.5))+
  guides(fill=guide_legend("Phylum"))+
  theme_few()

ra_plot=barplot_RA(biom$RA.Otus,tax = biom$taxon,meta = meta,category = "acetate",top = 5)

ggplot(data = ra_plot$top_long,aes(x=factor(label,levels = c("-1","0","0.5","1")),y=value,fill=variable))+ # this reorders x     axis values
  geom_bar(stat="identity")+
  scale_x_discrete("Acetate")+ # Labels categorical labels on x axis
  ylab("Relative Abundance (%)")+ 
  theme(legend.title=element_text(),legend.position="right",plot.title = element_text(hjust=0.5))+
  guides(fill=guide_legend("Phylum"))+
  theme_few()
```

```{r nmds plots}
veg=vegan_wrapper(biom$RA.Otus,meta = meta,category_1 = "temp",category_2 = "acetate")

veg$NMDS_plot+
    theme_few()+
    scale_shape_discrete(name="Acetate")+
    scale_colour_discrete(name = "Temperature")
    
# pull data out to better make plot
NMDS.data <- veg$NMDS_plot$data
NMDS.data <- NMDS.data %>% mutate_at(vars(category_1,category_2),funs(as.factor))

# https://gist.github.com/rmaia/5296401
shapes = c(21:24)
stroke = 1.5
 ggplot(data=NMDS.data, aes(x=MDS1, y=MDS2, color=category_1)) + geom_point(aes(shape=category_2),size=4,stroke=stroke)+
       scale_shape_manual(name = "Acetate",values=shapes)+
    scale_colour_discrete(name = "Temperature")
   #scale_colour_grey(name = "Temperature", start = 0.2, end = 0.8)
# zoom in 
   ggplot(data=NMDS.data, aes(x=MDS1, y=MDS2, color=category_1)) + geom_point(aes(shape=category_2),size=4,stroke=stroke)+
       scale_shape_manual(name = "Acetate",values=shapes)+
    scale_colour_discrete(name = "Temperature")+
   #scale_colour_grey(name = "Temperature", start = 0.2, end = 0.8)+
     coord_cartesian(xlim = c(.6, .9),ylim = c(-.2,-.1)) 
rm(shapes)
rm(stroke)
```

```{r ecological metrics}
# calc richnness, evenness, diversity each community
# diversity
div <- diversity(biom$RA.Otus, index = "shannon")
# richness
rich <- specnumber(biom$RA.Otus)
# evenness
even <- diversity(biom$RA.Otus, index = "shannon")/log(specnumber(biom$RA.Otus))
# all in one
rsm_hist <- data.frame(Label = names(div),diversity = div,richness = rich,evenness = even,row.names = NULL) %>% 
  left_join(meta,.) %>% 
  left_join(read_csv("rsm_hist.csv"),.) %>% #data from first script RSM_code.Rmd. Start building rsm_hist here in script.
  arrange(mfc)
```

```{r rsm of ecological metrics}
# without validation points, then stepped with AIC
dfVariable <- rsm_hist %>% select(mfc,type,temp,acetate,diversity,richness,evenness) %>% 
  filter(type!="validation") %>% 
  gather("variable","value",c(5:ncol(.))) %>% 
  group_by(variable) %>% 
  do(fitVariable = step(trace=0,lm(value ~ temp+acetate+I(temp^2)+I(acetate^2)+temp:acetate, data = .)))

# get the coefficients by group in a tidy data_frame
dfVariableCoef = tidy(dfVariable, fitVariable)
dfVariableCoef

# get the predictions by group in a tidy data_frame 
dfVariablePred = augment(dfVariable,fitVariable)
dfVariablePred

# get the summary statistics by group in a tidy data_frame
dfVariableSumm = glance(dfVariable, fitVariable)
dfVariableSumm


# get all significant models
dfVariableSumm %>% filter(p.value<=.05) %>% select(variable) %>% unique() %>% as.matrix() %>% match(.,dfVariable$variable) -> sig_models
dfVariable$variable[sig_models]

# loop it and provide titles - looks like it doesn't know what to do when acetate coefficients are undefined
alternativeFunction <- function(){
    newdat = data.frame(temp = seq(min(rsm_hist$temp),max(rsm_hist$temp),length.out = 100))
  newdat$pred = predict(model, newdata = newdat)
  plot(unlist(rsm_hist[dfVariable$variable[i]]) ~ temp, data = rsm_hist, main = paste(dfVariable$variable[i],"excluding validation points (AIC model)"),xlab="Temperature",ylab=dfVariable$variable[i])
  with(newdat, lines(x = temp, y = pred))
    }

for(i in sig_models){
  require(rsm)
  dfVariable$fitVariable[i] %>% unlist(recursive = FALSE) %>% structure(class = "lm") -> model
t <- try(
  persp(model, ~ temp + acetate, col = rainbow(50), contours = "colors", main = paste(dfVariable$variable[i],"excluding validation points (AIC model)")))
if("try-error" %in% class(t)) 
  alternativeFunction()
}

```
Also a good idea to try transforming data and doing RSM.
```{r}
rsm_hist <- rsm_hist %>% select(diversity,richness,evenness) %>% apply_transform %>% cbind(rsm_hist,.)

# without validation points, then stepped with AIC
dfVariable <- rsm_hist %>% select(mfc,type,temp,acetate,trans_diversity,trans_richness,trans_evenness) %>% 
  filter(type!="validation") %>% 
  gather("variable","value",c(5:ncol(.))) %>% 
  group_by(variable) %>% 
  do(fitVariable = step(trace=0,lm(value ~ temp+acetate+I(temp^2)+I(acetate^2)+temp:acetate, data = .)))

# get the coefficients by group in a tidy data_frame
dfVariableCoef = tidy(dfVariable, fitVariable)
dfVariableCoef


# # get the predictions by group in a tidy data_frame 
dfVariablePred = augment(dfVariable,fitVariable)
dfVariablePred

# get the summary statistics by group in a tidy data_frame
dfVariableSumm = glance(dfVariable, fitVariable)
dfVariableSumm


# get all significant models
dfVariableSumm %>% filter(p.value<=.05) %>% select(variable) %>% unique() %>% as.matrix() %>% match(.,dfVariable$variable) -> sig_models
dfVariable$variable[sig_models]

# This should probably be turned into a function in the functions.R script
# loop it and provide titles - looks like it doesn't know what to do when acetate coefficients are undefined
alternativeFunction <- function(){
    newdat = data.frame(temp = seq(min(rsm_hist$temp),max(rsm_hist$temp),length.out = 100))
  newdat$pred = predict(model, newdata = newdat)
  plot(unlist(rsm_hist[dfVariable$variable[i]]) ~ temp, data = rsm_hist, main = paste(dfVariable$variable[i],"excluding validation points (AIC model)"),xlab="Temperature",ylab=dfVariable$variable[i])
  with(newdat, lines(x = temp, y = pred))
    }

for(i in sig_models){
  dfVariable$fitVariable[i] %>% unlist(recursive = FALSE) %>% structure(class = "lm") -> model
t <- try(
  persp(model, ~ temp + acetate, col = rainbow(50), contours = "colors", main = paste(dfVariable$variable[i],"excluding validation points (AIC model)")))
if("try-error" %in% class(t)) 
  alternativeFunction()
}

```
#OTUs vs Clusters
```{r Co-occurrence across all OTUs, message = FALSE, warning=FALSE,fig.width=9,fig.height=9}
  
  # This is an area of networks where we study OTU's in the same environment that "co-occur". First we will filter data by co-occurence. Minimum recommended is 20%.
  
  #filter OTU's to 50% presence in all samples
  biom_fil=cooccur_filter(RA=biom$RA.Otus,co_per=0.5)
  
  #run co-occurence. Taxon can be excluded and identified later if desired.
  biom_netw=cooccurrence(biom_fil,taxon = biom$taxon)
  
  #try plotting the data
  # plot.igraph(biom_netw$netw) # standard plot function for igraph package
  # tkplot(biom_netw$netw) #interactive plot viewer
  # rglplot(biom_netw$netw) #experimental 3d plot viewer
  
  #Or by using ggnet2 (https://briatte.github.io/ggnet/)
  packages(network)
  packages(sna)
  packages(RColorBrewer)
  packages(intergraph)
  
  # Visualize
  set.seed(101)
  (network_plot <- ggnet2(biom_netw$netw, size = 6, color = "tomato", label=T, label.size=2))
```
Cluster network plot
```{r plot community networks,fig.width=12,fig.height=9}
# infomap community detection. Try the different detection algorithms to understand how different your niches might be broken up
biom_info=infomap.community(biom_netw$netw)


#now add some color to your previous plot (Josh's Method)
  #plot(biom_netw$netw,vertex.color=as.factor(biom_info$membership))
  
#Create a custom color scale
n <- max(biom_info$membership)
library(RColorBrewer)
myColors <- colorRampPalette(brewer.pal(12, "Paired"))(n)
names(myColors) <- levels(as.factor(biom_info$membership))
colScale <- scale_colour_manual(name = "Cluster",values = myColors)

#Plot with ggnet
nodesize <- 8
set.seed(101)
(plot_ggnet <- ggnet2(biom_netw$netw, size = nodesize, 
                      node.color = as.factor(biom_info$membership), 
                      color.legend = "Cluster",
                      label=T, 
                      label.size=2)+
    colScale+
    guides(colour = guide_legend(override.aes = list(size=nodesize))))
    

``` 
#Samples vs KOs
```{r heat maps of KOs across treatments}
# get metagenome data from PICRUST
predicted_metagenome = read.biom("C:/Users/Ryan/Google Drive/UC Davis/Dissertation for PhD/data/Exp23/metagenome.tab",new=T,metagenome=T)
# predicted_metagenome$RA.genes[1:10,1:10] # View data # Note this gives relative abundance in percent from 0 to 100

# Modify data to align it               -
meta$Label
rownames(predicted_metagenome$RA.genes) # see if row names match up in same order
predicted_metagenome$RA.genes <- predicted_metagenome$RA.genes[match(meta$Label,rownames(predicted_metagenome$RA.genes)),]
rownames(predicted_metagenome$RA.genes)

# Aggregate genes by sample (try first just by temp)
genes_agg = aggregate(predicted_metagenome$RA.genes,by=list(meta$temp),FUN=mean)
# Convert wide format into long format table
genes_melt=melt(genes_agg,id="Group.1")
# Filter by whatever genes of interest
kos_exoelectrogenic <- read.csv("data/kos_exoelectrogenic.csv",header=FALSE,fileEncoding="UTF-8-BOM")
genes_filtered <- genes_melt[which(genes_melt$variable %in% unlist(kos_exoelectrogenic)),]
# Make heat map
genes_filtered %>% 
  ggplot(data=.,aes(x=Group.1,y=variable,fill=value))+
  geom_tile()+
  scale_fill_gradient2("Gene relative abundance \n",guide="colourbar",high="#FF5733",low="#FDFEFE")+
  xlab("Temperature")+
  ylab("KEGG Ortholog")
# Now what if we rescale the data row-wise?
genes_filtered %>% 
  group_by(variable) %>% 
  mutate(rescale = scales::rescale(value)) %>% # this rescales each value 0:1 by variable
  ggplot(data=.,aes(x=Group.1,y=variable,fill=rescale))+
  geom_tile()+
  scale_fill_gradient2("Scaled gene relative abundance \n",guide="colourbar",high="#FF5733",low="#FDFEFE")+
  xlab("Temperature")+
  ylab("KEGG Ortholog")

# Aggregate genes by sample (try by acetate)
genes_agg = aggregate(predicted_metagenome$RA.genes,by=list(meta$acetate),FUN=mean)
# Convert wide format into long format table
genes_melt=melt(genes_agg,id="Group.1")
# Filter by whatever genes of interest
kos_exoelectrogenic <- read.csv("data/kos_exoelectrogenic.csv",header=FALSE,fileEncoding="UTF-8-BOM")
genes_filtered <- genes_melt[which(genes_melt$variable %in% unlist(kos_exoelectrogenic)),]
# Make heat map
genes_filtered %>% 
  ggplot(data=.,aes(x=Group.1,y=variable,fill=value))+
  geom_tile()+
  scale_fill_gradient2("Gene relative abundance \n",guide="colourbar",high="#FF5733",low="#FDFEFE")+
  xlab("Acetate")+
  ylab("KEGG Ortholog")
# Now what if we rescale the data row-wise?
genes_filtered %>% 
  group_by(variable) %>% 
  mutate(rescale = scales::rescale(value)) %>% # this rescales each value 0:1 by variable
  ggplot(data=.,aes(x=Group.1,y=variable,fill=rescale))+
  geom_tile()+
  scale_fill_gradient2("Scaled gene relative abundance \n",guide="colourbar",high="#FF5733",low="#FDFEFE")+
  xlab("Acetate")+
  ylab("KEGG Ortholog")

```
```{r RSM of each exoelectrogenic KO to temp and acetate}
# Aggregate genes by treatment or mfc in this case
genes_agg <- predicted_metagenome$RA.genes %>% mutate(Label = rownames(.)) %>% select(Label,everything())
# Convert wide format into long format table
genes_melt <- genes_agg %>% gather(.,key = "variable",value = "value",-Label)
# Filter by whatever genes of interest
genes_filtered <- genes_melt[which(genes_melt$variable %in% unlist(kos_exoelectrogenic)),]
#append to rsm_hist
rsm_hist <- genes_filtered %>% spread(variable,value) %>% left_join(rsm_hist,.,by="Label")
#apply transform
rsm_hist <- rsm_hist %>% select(starts_with("K")) %>% apply_transform %>% bind_cols(rsm_hist,.)
#run rsm and visualize
source("R/functions.R")
rsm_object <- rsm_and_visualize(data = rsm_hist,response_cols = rsm_hist %>% select(starts_with("trans_K")) %>% colnames)
rsm_object

# how about for the average of all the exoelectrogenic KOs?
rsm_hist <- rsm_hist %>% select(mfc,starts_with("trans_K")) %>% gather(trans_KO,value,-mfc) %>% group_by(mfc) %>% summarise(trans_KO_average = mean(value)) %>% right_join(rsm_hist,.)
rsm_object <- rsm_and_visualize(data = rsm_hist,response_cols = "trans_KO_average")
rsm_object

```
#Clusters vs KOs
```{r aggregate KOs by network cluster}
# Read in metagenome contributions table (takes 2 min to read)
a <- proc.time()
ko_metagenome_contributions <- read.table("C:/Users/Ryan/Google Drive/UC Davis/Dissertation for PhD/data/Exp23/metagenome_contributions.tab", header = T, sep = "\t", fill=TRUE)
proc.time()-a
# chop off unnecessary phylogeny
ko_metagenome_contributions <- ko_metagenome_contributions %>% select(-c(Kingdom,Phylum,Class,Order,Family,Genus,Species))
# which quantitative column is of interest? GeneCountPerGenome because it's the only one that doesn't involve relative abundance normalized to sample

# create lookup table
otu_cluster_lookup <- data.frame("OTU" = biom_info$names,"Cluster" = biom_info$membership)
# Lookup cluster number from lookup table
ko_metagenome_contributions$Cluster <- otu_cluster_lookup[match(ko_metagenome_contributions$OTU,otu_cluster_lookup$OTU),"Cluster"]

#filter out NA clusters
filt_ko_metagenome_contributions <- ko_metagenome_contributions[!is.na(ko_metagenome_contributions$Cluster),]
#remove unnecessary columns
filt_ko_metagenome_contributions <- filt_ko_metagenome_contributions %>% select(-c(OTUAbundanceInSample,CountContributedByOTU,ContributionPercentOfSample,ContributionPercentOfAllSamples))
head(filt_ko_metagenome_contributions)

# within each cluster, sum GeneCountPerGenome by Gene
filt_ko_metagenome_contributions <- aggregate(GeneCountPerGenome~Gene+Cluster,filt_ko_metagenome_contributions,sum)

# calc relative abundance of KOs within total KO GeneCountPerGenome in cluster
filt_ko_metagenome_contributions <- filt_ko_metagenome_contributions %>% group_by(Cluster) %>% mutate(relAbundByCluster = GeneCountPerGenome / sum(GeneCountPerGenome))

# filter by target KOs
filt_ko_metagenome_contributions <- filt_ko_metagenome_contributions[which(filt_ko_metagenome_contributions$Gene %in% unlist(kos_exoelectrogenic)),]

# tidy up names
colnames(filt_ko_metagenome_contributions) <- c("Gene","Cluster","GeneCount","relAbund")


#heat map
filt_ko_metagenome_contributions %>%  
    ggplot(data=.,aes(x=factor(Cluster),y=Gene,fill=relAbund))+
  geom_tile()+
  scale_fill_gradient2("Gene relative abundance \n",guide="colourbar",high="#FF5733",low="#FDFEFE")+
  xlab("Cluster")+
  ylab("KEGG Ortholog")+
  theme_few()+
 scale_x_discrete(expand = c(0,0))+
 scale_y_discrete(expand = c(0,0)) +
 coord_fixed(ratio=1)

# Rescale 0 to 1 for heat map
filt_ko_metagenome_contributions %>%  
  group_by(Gene)%>% 
  mutate(rescale = scales::rescale(relAbund)) %>% 
  ggplot(data=.,aes(x=factor(Cluster),y=Gene,fill=rescale))+
  geom_tile()+
  scale_fill_gradient2("Rescaled gene relative abundance \n",guide="colourbar",high="#FF5733",low="#FDFEFE")+
  xlab("Cluster")+
  ylab("KEGG Ortholog")+
  theme_few()+
 scale_x_discrete(expand = c(0,0))+
 scale_y_discrete(expand = c(0,0)) +
 coord_fixed(ratio=1)
```
```{r anova on which clusters have highest proportion of exoelectrogenic KOs}
filt_ko_metagenome_contributions %>%  select(-GeneCount) %>% 
  spread(Gene,relAbund) %>% arrange(Cluster) -> data
data %>% .[,-1] %>% apply_transform %>% bind_cols(data[,1],.) -> data

#plot
data %>% gather(key = "KO",value = "value",-Cluster) %>% 
ggplot(., aes(x=factor(Cluster), y = value)) +
    geom_boxplot(fill = "grey80", colour = "blue") +
    scale_x_discrete() + xlab("Cluster") +
    ylab("Transformed KO relative abundance") +
    theme_few()
#anova
data %>% 
  gather(key = "KO",value = "value",-Cluster) %>% 
  aov(value~factor(Cluster),data = .) %>% 
  summary
data %>% 
  gather(key = "KO",value = "value",-Cluster) %>% 
  aov(value~factor(Cluster),data = .) %>% 
  TukeyHSD -> Tukey
  Tukey$`factor(Cluster)`[Tukey$`factor(Cluster)`[,4]<.2,]

```
#Cluster vs Sample
```{r}
#assign clusters of interest
clusofinterest <- Tukey$`factor(Cluster)`[Tukey$`factor(Cluster)`[,4]<.2,] %>% rownames %>% as_tibble %>% separate(value,c("Cluster1","Cluster2")) %>% unlist %>% as.numeric %>% table %>% as_tibble %>% filter(n>1) %>% .[,1] %>% unlist %>% as.numeric

# make a long form table we can aggregate
ix <- which(colnames(biom$biom_tab) %in% c("Kingdom","Phylum","Class","Order","Family","Genus","Species")) # remove taxonomy info
data <- biom$biom_tab[,-ix] 
data <- t(data)
data <- melt(data)
colnames(data) <- c("Label","OTU","Reads")
#assign clusters to OTUs
data$Cluster <- otu_cluster_lookup[match(data$OTU,otu_cluster_lookup$OTU),"Cluster"]
# aggregate reads by label and cluster
agg_data <- aggregate(data$Reads,
    by = list(Label = data$Label, Cluster = data$Cluster),
    FUN = function(x) c(sum = sum(x)))
colnames(agg_data) <- c("Label","Cluster","Cluster_reads_in_label")
# aggregate reads by label
agg_data_label <- aggregate(data$Reads,
    by = list(Label = data$Label),
    FUN = function(x) c(sum = sum(x)))
# attach to agg_data
agg_data$label_reads <- agg_data_label[match(agg_data$Label,agg_data_label$Label),"x"]
# RA
agg_data$RA_cluster_in_label <- 100*agg_data$Cluster_reads_in_label/agg_data$label_reads

#reformat and attach to rsm_hist
agg_data %>% select(Label,Cluster,RA_cluster_in_label) %>% spread(key = Cluster,value = RA_cluster_in_label) -> data
colnames(data) %>% paste("cluster_",.,sep = "")->name_vector
"Label"->name_vector[1]
colnames(data) <- name_vector
data %>% left_join(rsm_hist,.,by = "Label") -> rsm_hist
data %>% .[,-1] %>% apply_transform %>% bind_cols(as_tibble(data$Label),.) %>% dplyr::rename("Label" = value) %>% left_join(rsm_hist,.,by = "Label") -> rsm_hist

#run rsm and visualize
source("R/functions.R")
rsm_object <- rsm_and_visualize(data = rsm_hist,response_cols = rsm_hist %>% select(starts_with("cluster")) %>% colnames)
rsm_object

#on transformed
rsm_object <- rsm_and_visualize(data = rsm_hist,response_cols = rsm_hist %>% select(starts_with("trans_cluster")) %>% colnames)
rsm_object
```



#KOs vs OTUs
```{r}
clusofinterest
# get data
data <- ko_metagenome_contributions
# filter by target OTUs
target_otus <- otu_cluster_lookup %>% filter(Cluster == clusofinterest) %>% select(OTU) %>% unlist %>% droplevels()
data <- droplevels(data[which(data$OTU %in% target_otus),])
data$OTU <- factor(data$OTU)
str(data)

# filter to one sample to get whole genome for each OTU (each gene should appear only once for each OTU)

data <- data %>% group_by(OTU) %>% distinct(Gene,.keep_all = TRUE)
# chop off unneccesary columns
data <- data %>% select(Gene,Sample,OTU,GeneCountPerGenome)

# calc relative abundance of KOs within total KO GeneCountPerGenome in OTU
data <- data %>% group_by(OTU) %>% mutate(relAbundByOTU = GeneCountPerGenome / sum(GeneCountPerGenome))

# filter by target KOs
data <- droplevels(data[which(data$Gene %in% unlist(kos_exoelectrogenic)),])

# Rescale 0 to 1 for heat map (decided not to use this time)
data <- data %>% group_by(Gene) %>% mutate(rescale = scales::rescale(relAbundByOTU))

# plot heat map
ggplot(data=data,aes(x=factor(OTU),y=Gene,fill=relAbundByOTU))+
  geom_tile()+
  scale_fill_gradient2("Gene relative abundance \n",guide="colourbar",high="#FF0000FF",mid="#FFFFD5FF")+
  xlab("OTU")+
  ylab("KEGG Ortholog")+
  scale_x_discrete(expand = c(0,0))+
 scale_y_discrete(expand = c(0,0)) +
 coord_fixed(ratio=1)
# can do anova on transformed KOs if needed

```
# Multivariate Analysis
```{r, fig.height=11, fig.width=11}

# add in phyla
phyla <- barplot_RA(biom$RA.Otus,tax = biom$taxon,meta = meta,category = "mfc",top = 5)$top_long %>% 
  spread(key = variable,value = value) %>% dplyr::rename("mfc" = label)

target_phyla <- phyla %>% select(-mfc) %>% colnames
phyla$mfc <- as.numeric(as.character(phyla$mfc))
phyla %>% select(-mfc) %>% apply_transform %>% bind_cols(phyla,.) %>% right_join(rsm_hist,.,by = "mfc") %>% arrange(mfc) -> rsm_hist

data <- rsm_hist %>% select(starts_with("trans"))

# https://cran.r-project.org/web/packages/corrplot/vignettes/corrplot-intro.html
packages(psych)
corr.test(data,adjust = "none") -> trans_corr
#non-adjusted p value matrix is symmetric, BH is not
p.adjust(trans_corr$p,method = "BH") %>% matrix(.,nrow = nrow(trans_corr$r),ncol = ncol(trans_corr$r)) -> trans_corr$p
packages(corrplot)
packages(RColorBrewer)
#make diagonal excluded
diag(trans_corr$p) <- 1
corrplot(trans_corr$r, type = "lower", method = "ellipse",order = "hclust", p.mat = trans_corr$p, sig.level = 0.05, insig = "blank",tl.col = "black", tl.srt = 45)
```
#Next steps from Chris
At this point, I think the main task is to connect specific OTUs, preferably with genus resolution, to expected exoelectrogenic activity via KO content or the scientific literature. Given the prominence of Thermotogae in the high temperature treatment, we should also comb the community for hydrogenotrophic microbes that may have syntrophic activity with acetate-oxidizing exoelectrogens.


Which OTUs were enriched in Cluster 5, our cluster that showed high levels of exoelectrogenic genes and were also enriched at high temperatures?
```{r}
target_otus
target_otus %>% as.character %>% as.numeric %>% tibble(otus=.) %>% left_join(.,biom$taxon)

df <- biom$biom_tab%>% select(-match(taxaNames,names(.))) %>% rownames_to_column(var="OTU") %>% filter(OTU %in% target_otus) %>% gather(Label,reads,-OTU) %>% left_join(.,rsm_hist %>% select(Label,temp,acetate),by="Label")

df %>% ggplot(data = ., aes(x = "", y = reads, fill = OTU )) + 
    geom_bar(stat = "identity", position = position_fill(),width=1)+
  coord_polar("y", start=0)+
  theme_few()+
  facet_wrap(~ temp)+
  theme(axis.title.x = element_blank(),
    axis.title.y = element_blank())+
  ggtitle("Cluster 5 OTUs")
   
#chi square for cluster 5
# https://stats.stackexchange.com/questions/2391/what-is-the-relationship-between-a-chi-squared-test-and-test-of-equal-proportion
mydata <- ko_metagenome_contributions
# chop off unneccesary columns
mydata[5:8] <- list(NULL)

# fix labels
names(mydata)[names(mydata) == 'Sample'] <- 'Label'
mydata$Label = paste("X",mydata$Label,sep = "")
mydata$Label = gsub("-",".",mydata$Label)
  
# filter for anodes
mydata <- mydata %>% filter(Label %in% meta$Label)

# make column for Cluster5 yes/no
mydata$Cluster5 <- ifelse(mydata$Cluster == 5,TRUE,FALSE)  

# make column for exoelectrogenic KO yes/no
mydata$exoelectrogenic_ko <- ifelse(mydata$Gene %in% unlist(kos_exoelectrogenic),TRUE,FALSE)  

# note that data is in the form of counts
# make contingency table
ctable <- xtabs(GeneCountPerGenome ~ Cluster5+exoelectrogenic_ko,data=mydata)
head(ctable)
# visualize
packages(vcd)
mosaic(ctable,shade=T,legend=T)
# get proportions
prop.table(ctable,1)
chisq.test(ctable)
```
Looks like two of the three OTUs in cluster 5 are Thermotogae.

Next, let's downselect for Thermotogae and other phyla and fit it to a response surface.
```{r}
response_cols = c("trans_Thermotogae","trans_Proteobacteria","trans_Firmicutes","trans_Bacteroidetes","trans_Tenericutes","trans_Other")
rsm_object <- rsm_and_visualize(data = rsm_hist,response_cols = response_cols)
rsm_object
```
Looks like all Thermotogae are highly enriched at higher temperatures. In fact, all 5 major phyla fit to our response surface. Bacteroidetes decreases at high temps, as does Proteobacteria, as does Tenericutes. What about the Firmicutes? trans_Firmicutes modeled to our response surface fits only to temperature, and has a global maximum around coded temp = 0.5.

What if we group our metagenome by phyla, then look at a heat map of exoelectrogenic KOs across phyla?
```{r}
data <- ko_metagenome_contributions %>% select(-c(OTUAbundanceInSample,CountContributedByOTU,ContributionPercentOfSample,ContributionPercentOfAllSamples))

#attach phylogeny
data <- data %>% inner_join(biom$taxon,by = c("OTU" = "otus"))

# within each phylum, sum GeneCountPerGenome by Gene, then calculate that gene's relative abundance
data %>% group_by(Phylum,Gene) %>% summarise(geneCount = sum(GeneCountPerGenome)) %>% group_by(Phylum) %>% mutate(relAbundByPhylum = geneCount / sum(geneCount)) %>% 
# filter by target KOs
filter(Gene %in% unlist(kos_exoelectrogenic)) -> data
  
  ggplot(data=data,aes(x=Phylum,y=Gene,fill=relAbundByPhylum))+
  geom_tile()+
  scale_fill_gradient2("Gene relative abundance \n",guide="colourbar",high="#FF5733",low="#FDFEFE")+
  xlab("Phylum")+
  ylab("KEGG Ortholog")+
  theme_few()+
 scale_x_discrete(expand = c(0,0))+
 scale_y_discrete(expand = c(0,0)) +
 coord_fixed(ratio=1)+
theme(axis.text.x = element_text(angle = 60, hjust = 1))
  
# Rescale 0 to 1 for heat map 
data %>%
  group_by(Gene)%>%
  mutate(rescale = scales::rescale(relAbundByPhylum)) %>%
  ggplot(data=.,aes(x=Phylum,y=Gene,fill=rescale))+
  geom_tile()+
  scale_fill_gradient2("Rescaled gene relative abundance \n",guide="colourbar",high="#FF5733",low="#FDFEFE")+
  xlab("Cluster")+
  ylab("KEGG Ortholog")+
  theme_few()+
 scale_x_discrete(expand = c(0,0))+
 scale_y_discrete(expand = c(0,0)) +
 coord_fixed(ratio=1)+
theme(axis.text.x = element_text(angle = 60, hjust = 1))
```
Thermotogae does greatly outweigh the other phyla in terms of relative abundance of the KEGG Ortholog K03406.

What happens if we look for OTUs with the highest counts/relative abundances of exoelectrogenic KOs in our high temperature samples?
This is total counts of exoelectrogenic genes per genome of each OTU, downselected for high temp samples and only exoelectrogenic genes. Two of the top three are Thermotogae.
```{r}
data <- ko_metagenome_contributions 

(data2 <- data %>% select(-c(OTUAbundanceInSample,CountContributedByOTU,ContributionPercentOfSample,ContributionPercentOfAllSamples)) %>% inner_join(biom$taxon,by = c("OTU" = "otus")) %>% mutate_at(vars(Sample),funs(paste("X",.,sep=""))) %>% filter(Sample %in% meta$Label[meta$temp==1]) %>% filter(Gene %in% unlist(kos_exoelectrogenic)) %>% group_by(OTU) %>% summarise(exoGeneCountPerGenome = sum(GeneCountPerGenome)) %>% arrange(desc(exoGeneCountPerGenome)) %>% inner_join(biom$taxon,by = c("OTU" = "otus")))
#what if we insert the number of times each OTU occurred in our high temp samples, and multiply?
biom$biom_tab %>% select(meta$Label[meta$temp==1]) %>% rownames_to_column(var = "OTU") %>% gather(Label,genomes_per_OTU,-OTU) %>% group_by(OTU) %>% summarise(genomes_per_OTU_high_temp = sum(genomes_per_OTU)) %>% mutate_at(vars(OTU),funs(as.numeric)) %>% left_join(data2,.,by="OTU") %>% select(OTU,exoGeneCountPerGenome,genomes_per_OTU_high_temp,everything()) %>% mutate(totalExoGenesHighTemp = exoGeneCountPerGenome*genomes_per_OTU_high_temp) %>% arrange(desc(totalExoGenesHighTemp)) %>% select(OTU,exoGeneCountPerGenome,genomes_per_OTU_high_temp,totalExoGenesHighTemp,everything())
# proves each gene-sample-otu combo appears only once, indicating the metagenome contributions are not multiplied by OTU count data
# data %>% select(-c(OTUAbundanceInSample,CountContributedByOTU,ContributionPercentOfSample,ContributionPercentOfAllSamples)) %>% mutate(combo = paste(Gene,Sample,OTU)) %>% select(combo) -> data2
# data2 %>% group_by(combo) %>% filter(n()>1)

#What about relative abundance of exo KOs out of all KOs for each specific OTU at high temps?
ko_metagenome_contributions %>% select(-c(OTUAbundanceInSample,CountContributedByOTU,ContributionPercentOfSample,ContributionPercentOfAllSamples)) %>% mutate_at(vars(Sample),funs(paste("X",.,sep=""))) %>% filter(Sample %in% meta$Label[meta$temp==1]) %>% 
#deduplicate gene-OTU combos across multiple samples
group_by(OTU) %>% distinct(Gene,.keep_all = TRUE) %>% 
#mutate rel abun
  group_by(OTU) %>% mutate(geneRelAbundPerGenome = GeneCountPerGenome/sum(GeneCountPerGenome)) %>% filter(Gene %in% unlist(kos_exoelectrogenic)) %>% ungroup %>% group_by(OTU) %>% summarise(avgExoGeneRelAbundPerGenome = mean(geneRelAbundPerGenome)) %>% arrange(desc(avgExoGeneRelAbundPerGenome)) %>% 
  #attach taxonomy
  left_join(biom$taxon,by = c("OTU" = "otus"))
  

```
This shows that at high temperatures, Thermotoga have both high counts of exoelectrogenic genes per genome as well as the highest total exoelectrogenic gene counts when gene counts per genome are multiplied by their OTU abundance. Finally, Thermotoga have the highest average exoelectrogenic gene relative abundance per genome.

What would the hydrogenotroph be? A firmicute since that was the other main phyla in the high temperature MFCs? The other OTU in Cluster 5?
Lit review on firmicutes and hydrogenotrophy - the firmicutes Clostridia and Bacilli have lots of EEG's
Clostridia are anaerobes, Bacilli are facultative


```{r}
#Cluster 5 symbiont and hydrogenotrophy?
target_otus %>% as.character %>% as.numeric %>% tibble(otus=.) %>% left_join(.,biom$taxon)

```
https://microbewiki.kenyon.edu/index.php/Bacteroides
Bacteroides are gram-negative, nonsporeforming, anaerobic, and rod-shaped bacteria. They have an outer membrane, a peptidoglycan layer, and a cytoplasmic membrane. The main by-products of their anaerobic respiration are acetic acid, iso valeric acid, and succinic acid. They are involved in many important metabolic activities in the human colon including fermentation of carbohydrates, utilization of nitrogenous substances, and biotransformation of bile acids and other steroids.
-> This group is possibly symbiotic with Thermotoga.


Methanogenesis is limited to archaea.
Can I detect Archaea? Only k_Bacteria listed in my taxonomy file. But I downselected for MFC anodes. Any in sludge?  
```{r}
fullBiom <- read.biom("data/tsv_otu_table.txt")
fullBiom %>% summary
fullBiom$biom_tab %>% group_by(Kingdom) %>% summarise(n())
fullBiom$biom_tab %>% filter(Kingdom == "k__Archaea") %>% gather(key,value,-c(Kingdom,Phylum,Class,Order,Family,Genus,Species)) %>% filter(value>0) %>% select(key,value,everything())

#Is there evidence that our lack of archaea is a function of binned otus?
dfArchTF <- fullBiom$biom_tab %>% filter(Kingdom == "k__Archaea") %>% gather(Label,value,-c(Kingdom,Phylum,Class,Order,Family,Genus,Species)) %>% select(Label,value) %>% mutate(archTF = value>1)

taxaNames <- c("Kingdom","Phylum","Class","Order","Family","Genus","Species")

df <- fullBiom$biom_tab %>% select(-match(taxaNames,names(.))) %>% rownames_to_column(var="OTU") %>% gather(Label,value,-OTU) %>% group_by(Label) %>% summarise(totalBins = sum(value)) %>% left_join(.,dfArchTF,by="Label")
df
#binomial logistic regression
df %>% ggplot(aes(x=totalBins,y=value))+geom_point()+ylab("archaeal bins")+xlab("total bins per sample")

model <- glm(archTF ~ totalBins,family=binomial(link='logit'),data=df)
model %>% summary
```
So Archaea were detected, but not on the anode biofilm. They were detected in small amounts in the sludge inoculum, the bulk liquid, and one of the air cathodes.

Are there any archaea in the analysis from RTL? Yes, exact same counts on those exact same samples.
Also, RTL indicates my highly enriched thermotoga are Defluvitoga tunisiensis.

Any hydrogenotroph KOs on KEGG?
http://www.genome.jp/kegg-bin/get_htext#B12 was searched for hydrogenases, which all have EC numbers starting with 1.12.
```{r}
hydrogenaseKOdf <- read_csv("data/hydrogenaseKOs.csv",col_names = FALSE)
hydrogenaseKOdf <- hydrogenaseKOdf %>% unite(.,"description",X2:X9,sep = " ") %>% dplyr::rename(KO = X1)
hydrogenaseKO <- hydrogenaseKOdf %>% select(KO) %>% unlist %>% sort
hydrogenaseKO
```
Were any of these KOs in my anode metagenome?
```{r}
targetHydrogenaseKO <- predicted_metagenome$RA.genes %>% colnames %>% .[which(. %in% hydrogenaseKO)] %>% sort
targetHydrogenaseKO
length(targetHydrogenaseKO)
data <- predicted_metagenome$RA.genes %>% select(targetHydrogenaseKO) %>% rownames_to_column(var = "Label") %>%   #remove columns with all zeros
  gather(KO,RA,-Label) %>% filter(RA>0)
data
data %>% lm(RA~KO,data=.) %>% summary
data %>% aov(RA~KO,data=.) %>% TukeyHSD -> Tukey
Tukey$KO[Tukey$KO[,4]<.05,]
data %>% ggplot(aes(x=KO,y=RA))+geom_boxplot()+xlab("Hydrogenase KO")+ylab("RA across all samples (%)")

#Fit to our response surface?
#First each KO - spread and replace NA with zero
rsm_hist <- data %>% spread(.,KO,RA) %>% mutate_all(funs(replace(., is.na(.), 0))) %>% mutate_at(vars(starts_with("K")),funs(trans = mutate_transform(.))) %>% left_join(rsm_hist,.,by = "Label")


targetHydrogenaseKO[which(targetHydrogenaseKO %in% colnames(rsm_hist))]->targetHydrogenaseKO
targetHydrogenaseKO->response_cols
rsm_object <- rsm_and_visualize(data = rsm_hist,response_cols = response_cols)
rsm_object


rsm_hist %>% select(ends_with("trans")) %>% colnames -> response_cols
rsm_object <- rsm_and_visualize(data = rsm_hist,response_cols = response_cols)
rsm_object

# try with average hydrogenase KO content
rsm_hist %>% select(Label,targetHydrogenaseKO) %>% gather(key,value,-Label) %>% group_by(Label) %>% summarise(avgHydrKO = mean(value)) %>% left_join(rsm_hist,.,by = "Label") -> rsm_hist

rsm_hist %>% select(Label,ends_with("trans")) %>% gather(key,value,-Label) %>% group_by(Label) %>% summarise(avgHydrKO_trans = mean(value)) %>% left_join(rsm_hist,.,by = "Label") -> rsm_hist

rsm_hist %>% select(starts_with("avgHydr")) %>% colnames -> response_cols
rsm_object <- rsm_and_visualize(data = rsm_hist,response_cols = response_cols)
rsm_object


```
Interesting - looks like K00436, a hydrogenase KO, was found extensively across all samples. The average Hydrogenase KO content did not fit to our response surface. K00441 fits to our response surface. K00533 fits to our response surface as well. Notably, K00533 was enriched in our high temperature samples. Next question: which clusters and OTUs contain the highest levels of K00533?
```{r screen OTUs for target Hydrogenase KOs}
# get metagenome contributions from high temp samples
data <- ko_metagenome_contributions %>% select(-c(OTUAbundanceInSample,CountContributedByOTU,ContributionPercentOfSample,ContributionPercentOfAllSamples)) %>% mutate_at(vars(Sample),funs(paste("X",.,sep=""))) %>% filter(Sample %in% meta$Label[meta$temp==1]) 
# filter for K00533 and sort high to low
data %>% filter(Gene == "K00533") # this gene contained in these samples from these OTUs
data %>% filter(Gene == "K00533") %>% select(OTU) %>% unique %>% left_join(.,biom$taxon,by = c("OTU"="otus"))
# sum by all target Hydrogenase KOs and sort
data2 <- data %>% filter(Gene %in% targetHydrogenaseKO) %>% filter(Sample %in% meta$Label[meta$temp==1]) %>% group_by(OTU) %>% distinct(Gene,.keep_all = TRUE) %>% summarise(hydrGeneCountPerGenome = sum(GeneCountPerGenome)) %>% arrange(desc(hydrGeneCountPerGenome)) %>% left_join(.,biom$taxon,by = c("OTU"="otus"))
data2
#what if we insert OTU counts and multiply to get total hydrogenotrophic genes at high temps?
(df <- biom$biom_tab %>% select(meta$Label[meta$temp==1]) %>% rownames_to_column(var = "OTU") %>% gather(Label,genomes_per_OTU,-OTU) %>% group_by(OTU) %>% summarise(genomes_per_OTU_high_temp = sum(genomes_per_OTU)) %>% mutate_at(vars(OTU),funs(as.numeric)) %>% left_join(data2,.,by="OTU") %>% select(OTU,hydrGeneCountPerGenome,genomes_per_OTU_high_temp,everything()) %>% mutate(totalHydrGenesHighTemp = hydrGeneCountPerGenome*genomes_per_OTU_high_temp) %>% arrange(desc(totalHydrGenesHighTemp)) %>% select(OTU,hydrGeneCountPerGenome,genomes_per_OTU_high_temp,totalHydrGenesHighTemp,everything()))

#take these hydrogenotrophic OTUs, average their relative abundance in each label, then fit to response surface
hydrOTU <-  df %>% select(OTU) %>% unlist %>% as.character
rsm_hist <- biom$RA.Otus %>% rownames_to_column(var = "Label") %>% gather(key,value,-Label) %>% filter(key %in% hydrOTU) %>% group_by(Label) %>% summarise(avgAllHydrOTU = mean(value)) %>% mutate(trans_avgAllHydrOTU = mutate_transform(avgAllHydrOTU)) %>% left_join(rsm_hist,.,by="Label")
#fit to response surface
rsm_hist %>% select(contains("avgAllHydrOTU")) %>% colnames -> response_cols
rsm_object <- rsm_and_visualize(data = rsm_hist,response_cols = response_cols)
rsm_object

rsm_object$dfVariable$fitVariable[2] %>% unlist(recursive = FALSE) %>% structure(class = "lm") -> model
newdat = data.frame(temp = seq(min(rsm_hist$temp),max(rsm_hist$temp),length.out = 100))
newdat$pred = predict(model, newdata = newdat)
plot(trans_avgAllHydrOTU ~ temp, data = rsm_hist, main = "trans_avgAllHydrOTU fit to temp (AIC model)")
with(newdat, lines(x = temp, y = pred))


```
Interesting. If we average the relative abundance of all the OTUs containing hydrogenotrophic genes, it fits to temperature but not acetate, and reaches saturation at higher temperatures.

Haloplasmatales is strictly anaerobic and halophilic and has an optimum temperature for growth of about 30 to 37°C and an optimum pH of about 7. Nitrate and nitrite are reduced; lactate is a fermentation product. (http://jb.asm.org/content/190/10/3580.full)

Maybe Halanaerobiaceae? Within the Halobacteroidaceae, a greater metabolic diversity is found, with some species displaying a homoacetogenic metabolism; growth by anaerobic respiration using different electron acceptors including nitrate, trimethylamine.-oxide, selenate, arsenate, or Fe(III); or chemolithoautotrophic growth on hydrogen and elemental sulfur.
```{r}
#what if we fit these OTUs with high numbers of hydrogenotrophic genes to our response surface?
rsm_temp <- rsm_hist %>% select(mfc,type,temp,acetate,Label)

targetHydrogenaseOTU <- data2 %>% select(OTU) %>% unlist %>% as.character
targetHydrogenaseOTU <- targetHydrogenaseOTU[targetHydrogenaseOTU %in% colnames(biom$RA.Otus)]
data <- biom$RA.Otus %>% select(targetHydrogenaseOTU) %>% rownames_to_column(var = "Label") %>% left_join(rsm_temp,.,by = "Label") %>% mutate_at(vars(-c(mfc,type,temp,acetate,Label)),funs(trans = mutate_transform))

data %>% select(ends_with("trans")) %>% colnames -> response_cols
rsm_object <- rsm_and_visualize(data = data,response_cols = response_cols)
rsm_object

# These are the hydrogenotrophic OTUs that fit our reponse surface
rsm_object$dfVariableSumm %>% filter(p.value<.05)
rsm_object$dfVariableSumm %>% filter(p.value<.05) %>% select(variable) %>% unlist %>% str_replace_all(.,"_trans","") %>% as.numeric %>% tibble(OTU=.) %>% left_join(.,biom$taxon,by = c("OTU"="otus"))
fitTargetHydrogenaseOTU <- rsm_object$dfVariableSumm %>% filter(p.value<.05) %>% select(variable) %>% unlist %>% str_replace_all(.,"_trans","")

data %>% select(temp,fitTargetHydrogenaseOTU) %>% gather(OTU,RA,-temp) %>% ggplot(aes(x=temp,y=RA))+geom_point()+facet_wrap(~OTU,scales="free")

#Did these target hydrogenase OTUs cluster in our network? Nope
fitTargetHydrogenaseOTU %>% as.numeric %>% match(.,otu_cluster_lookup$OTU)

#Did any of these target hydrogenase OTUs correlate with our trans_Thermotoga?
mycols <- rsm_object$dfVariableSumm %>% filter(p.value<.05) %>% select(variable) %>% unlist
corrdata <- data %>% select(Label,mycols) %>% left_join(.,rsm_hist %>% select(Label,trans_Thermotogae)) %>% select(-Label)

# https://cran.r-project.org/web/packages/corrplot/vignettes/corrplot-intro.html
packages(psych)
corr.test(corrdata,adjust = "none") -> trans_corr
#non-adjusted p value matrix is symmetric, BH is not
p.adjust(trans_corr$p,method = "BH") %>% matrix(.,nrow = nrow(trans_corr$r),ncol = ncol(trans_corr$r)) -> trans_corr$p
packages(corrplot)
packages(RColorBrewer)
#make diagonal excluded
diag(trans_corr$p) <- 1
corrplot(trans_corr$r, type = "lower", method = "ellipse",order = "hclust", p.mat = trans_corr$p, sig.level = 0.05, insig = "blank",tl.col = "black", tl.srt = 45)


```
These hydrogenotrophic OTUs dont' seem to correlate with Thermotogae, either Pearson or Spearman, although their reponse surfaces look similar.

Do none of the potentially hydrogenotrophic OTUs correlate with Thermotogae RA on their own?
```{r}
# get metagenome contributions from all samples
df <- ko_metagenome_contributions %>% select(-c(OTUAbundanceInSample,CountContributedByOTU,ContributionPercentOfSample,ContributionPercentOfAllSamples)) %>% mutate_at(vars(Sample),funs(paste("X",.,sep="")))

# filter for hydrogenase KOs
df <- df %>% filter(Gene %in% targetHydrogenaseKO)


corrdata <- biom$RA.Otus %>% rownames_to_column(var = "Label") %>% gather(key,value,-Label) %>% filter(key %in% (df %>% select(OTU) %>% unlist %>% unique)) %>% spread(key,value) %>% left_join(rsm_temp,.,by="Label") %>% left_join(rsm_hist %>% select(Label,trans_Thermotogae),.,by = "Label") %>% select(-Label,-mfc,-type,-temp,-acetate)

# https://cran.r-project.org/web/packages/corrplot/vignettes/corrplot-intro.html
packages(psych)
corr.test(corrdata,adjust = "none") -> trans_corr
#non-adjusted p value matrix is symmetric, BH is not
p.adjust(trans_corr$p,method = "BH") %>% matrix(.,nrow = nrow(trans_corr$r),ncol = ncol(trans_corr$r)) -> trans_corr$p
packages(corrplot)
packages(RColorBrewer)
#make diagonal excluded
diag(trans_corr$p) <- 1
corrplot(trans_corr$r, type = "lower", method = "ellipse",order = "hclust", p.mat = trans_corr$p, sig.level = 0.05, insig = "blank",tl.col = "black", tl.srt = 45)


# Are these 4 OTUs also the most abundant potential hydrogenotrophs in these communities? Yes, in our high temp samples.
biom$RA.Otus %>% rownames_to_column(var = "Label") %>% gather(key,value,-Label) %>% filter(key %in% (df %>% select(OTU) %>% unlist %>% unique)) %>% spread(key,value) %>% left_join(rsm_temp,.,by="Label") %>% select(-mfc,-type,-temp,-acetate) %>% gather(key,value,-Label) %>% mutate(log_RA = log10(value),color = key %in% c(1105176,153243,242302,784950)) %>% ggplot(aes(x=Label,y=log_RA,colour = color))+geom_point()+theme(axis.text.x = element_text(angle = 60, hjust = 1))+scale_colour_discrete(name="OTU fits significantly to temperature")+xlab("Sample")+ylab("Log 10 Relative Abundance")

pot_hydrogenotrophs <- df %>% select(OTU) %>% unlist %>% unique


# Do these 4 OTUs have the most total hydrogenotrophic genes in these communities?
df <- ko_metagenome_contributions %>% filter(Gene %in% targetHydrogenaseKO) %>% group_by(OTU) %>% distinct(Gene,.keep_all = TRUE) %>% summarise(hydrGeneCountPerGenome = sum(GeneCountPerGenome)) %>% arrange(desc(hydrGeneCountPerGenome))

biom$biom_tab %>% rownames_to_column(var = "OTU") %>% select(-match(taxaNames,names(.))) %>% gather(Label,genomes_per_OTU,-OTU) %>% mutate_at(vars(OTU,genomes_per_OTU),funs(as.numeric)) %>% filter(OTU %in% pot_hydrogenotrophs) %>% left_join(.,df,by="OTU") %>% mutate(totalHydrGenes = genomes_per_OTU*hydrGeneCountPerGenome) %>% mutate(log_totalHydrGenes = log10(totalHydrGenes)) %>% 
  mutate(color = OTU %in% c(1105176,153243,242302,784950)) %>% 
  ggplot(aes(x=Label,y=log_totalHydrGenes,colour = color))+
  geom_point()+
  theme(axis.text.x = element_text(angle = 60, hjust = 1))+
  scale_colour_discrete(name="Four Down Selected OTUs")
 
# can we do a line plot of %RA fit to temperature for each pot hydrogenotroph? Color by OTU name.
biom$RA.Otus %>% rownames_to_column(var = "Label") %>% gather(key,value,-Label) %>% filter(key %in% pot_hydrogenotrophs) %>% spread(key,value) %>% left_join(rsm_temp,.,by="Label") %>% select(-c(mfc,type,acetate,Label)) %>% gather(key,value,-temp) %>% group_by(temp,key) %>% summarise(avg = mean(value)) %>% mutate(log_avg = log10(avg)) %>% 
  ggplot(aes(x=temp, y=log_avg, group=key)) +
  geom_line(aes(color=key))+
  geom_point(aes(color=key))+
  xlab("Temperature")+
  ylab("Log 10 Relative Abundance")+
  scale_colour_discrete(name="OTU")+
  theme_few()
  
```


Were there more hydrogenotrophic KOs in Cluster 5, our one cluster enriched at higher temperatures?
```{r No hydrogenotrophic KOs in Cluster 5}
#No hydrogenotrophic KOs in Cluster 5
targetHydrogenaseKO
ko_metagenome_contributions %>% select(-c(OTUAbundanceInSample,CountContributedByOTU,ContributionPercentOfSample,ContributionPercentOfAllSamples)) %>% mutate_at(vars(Sample),funs(paste("X",.,sep=""))) %>% filter(Cluster == 5) %>% filter(Gene %in% targetHydrogenaseKO)
```
What if we average counts across all hydrogenotrophic OTUs?
Then avg hydrogenotroph OTU RA fits well to temperature, and correlates well with Thermotogae.
```{r}
rsm_hist <- data %>% select(Label,mycols) %>% gather(OTU,value,-Label) %>% group_by(Label) %>% summarise(trans_avgFitHydrOTU = mean(value)) %>% left_join(rsm_hist,.,by = "Label")

rsm_object <- rsm_and_visualize(data = rsm_hist,response_cols = "trans_avgFitHydrOTU")
rsm_object
# correlation between thermotoagae and trans_avgFitHydrOTU
packages(ggpubr)
rsm_hist %>% select(Label,temp,acetate,trans_avgFitHydrOTU) %>% left_join(rsm_hist %>% select(Label,trans_Thermotogae)) %>%  ggplot(aes(trans_Thermotogae,trans_avgFitHydrOTU))+geom_point()+geom_smooth(method="lm")+stat_cor()

#Total RA instead of avg RA?

data %>% select(Label,mycols) %>% gather(OTU,value,-Label) %>% group_by(Label) %>% summarise(trans_sumFitHydrOTU = sum(value)) %>% left_join(rsm_hist,.,by = "Label") %>% ggplot(aes(trans_Thermotogae,trans_sumFitHydrOTU))+geom_point()+geom_smooth(method="lm")+stat_cor()
```
#Final multivariate analysis
```{r, fig.height=11, fig.width=11}
df <- rsm_hist %>% select(contains("trans"))
# https://cran.r-project.org/web/packages/corrplot/vignettes/corrplot-intro.html
packages(psych)
corr.test(df,adjust = "none") -> trans_corr
#non-adjusted p value matrix is symmetric, BH is not
p.adjust(trans_corr$p,method = "BH") %>% matrix(.,nrow = nrow(trans_corr$r),ncol = ncol(trans_corr$r)) -> trans_corr$p
packages(corrplot)
packages(RColorBrewer)
#make diagonal excluded
diag(trans_corr$p) <- 1
corrplot(trans_corr$r, type = "lower", method = "ellipse",order = "hclust", p.mat = trans_corr$p, sig.level = 0.05, insig = "blank",tl.col = "black", tl.srt = 45)
```


#Modeling power based on biofilm properties
See if I can build a model for power metrics based on biofilm properties (structure and phylogeny)
```{r}
# make sure all relevant biofilm data has been added to rsm_hist
colNames <- rsm_hist %>% select(contains("trans")) %>% colnames
strMetrics <- c("trans_PI"     ,             "trans_SG"      ,           "trans_live_dead"  )
powerMetrics <- c("trans_charge_C","trans_ce","trans_delta_cod_mg_L"    ,  "trans_delta_cod_mg_L_day",  "trans_voltage",  "trans_cycle_5_energy_J"  ,  "trans_cumulative_energy_J")
phyMetrics <- colNames[!(colNames %in% strMetrics) & !(colNames %in% powerMetrics)]

#model
#note that step needs an explicit data frame in the model call
df <- rsm_hist %>% select(trans_cumulative_energy_J,strMetrics)
lm(trans_cumulative_energy_J ~ (.)^2, data = df) %>% step(trace = 0,.) %>% summary

df <- rsm_hist %>% select(trans_cumulative_energy_J,trans_avgAllHydrOTU,trans_avgFitHydrOTU)
lm(trans_cumulative_energy_J ~ (.)^2, data = df) %>% step(trace = 0,.) %>% summary

df <- rsm_hist %>% select(trans_cumulative_energy_J,trans_avgFitHydrOTU)
lm(trans_cumulative_energy_J ~ (.)^2, data = df) %>% step(trace = 0,.) %>% summary

df <- rsm_hist %>% select(trans_cumulative_energy_J,trans_Thermotogae,trans_avgFitHydrOTU)
lm(trans_cumulative_energy_J ~ (.)^2, data = df) %>% step(trace = 0,.) %>% summary

df <- rsm_hist %>% select(trans_cumulative_energy_J,strMetrics,trans_Thermotogae)
lm(trans_cumulative_energy_J ~ (.)^2, data = df) %>% step(trace = 0,.) %>% summary
#leaving at this for now, since cumulative energy has been our best dependent variable fitted to our response surface
```
Note that the best prediction of cumulative energy production is Thermotogae, not thermotogae combined with avgFitHydrOTU.
We get a decent model if we include all the structural metrics and Thermotogae relative abundance.

Plot residuals for selected model
```{r fig.height=9, fig.width=9}
df <- rsm_hist %>% select(trans_cumulative_energy_J,strMetrics,trans_Thermotogae)
model <- lm(trans_cumulative_energy_J ~ (.)^2, data = df) %>% step(trace = 0,.)
par(mfrow = c(2, 2))
plot(model)
#looks like some non-normality in the QQ plot. let's plot for each predictor and see what happens
# https://drsimonj.svbtle.com/visualising-residuals
# Obtain predicted and residual values
df$predicted <- predict(model)
df$residuals <- residuals(model)

# Create plot
df %>% 
  gather(key = "iv", value = "x", -trans_cumulative_energy_J, -predicted, -residuals) %>%
  ggplot(aes(x = x, y = trans_cumulative_energy_J)) +
  geom_segment(aes(xend = x, yend = predicted), alpha = .2) +
  geom_point(aes(color = residuals)) +
  scale_color_gradient2(low = "blue", mid = "white", high = "red") +
  guides(color = FALSE) +
  geom_point(aes(y = predicted), shape = 1) +
  facet_grid(~ iv, scales = "free") +
  theme_few()

#try again with squared terms?
df <- rsm_hist %>% select(trans_cumulative_energy_J,strMetrics,trans_Thermotogae)
model <- lm(trans_cumulative_energy_J ~ I(trans_SG^2)+trans_PI+trans_live_dead+trans_Thermotogae+(.)^2, data = df) %>% step(trace = 0,.)
summary(model)
par(mfrow = c(2, 2))
plot(model)

df <- rsm_hist %>% select(trans_cumulative_energy_J,strMetrics,trans_Thermotogae)
model <- lm(trans_cumulative_energy_J ~ I(trans_SG^2)+I(trans_PI^2)+trans_live_dead+trans_Thermotogae+(.)^2, data = df) %>% step(trace = 0,.)
summary(model)
par(mfrow = c(2, 2))
plot(model)

df <- rsm_hist %>% select(trans_cumulative_energy_J,strMetrics,trans_Thermotogae)
model <- lm(trans_cumulative_energy_J ~ I(trans_PI^2)+I(trans_live_dead^2)+trans_Thermotogae+(.)^2, data = df) %>% step(trace = 0,.)
summary(model)
par(mfrow = c(2, 2))
plot(model)

df <- rsm_hist %>% select(trans_cumulative_energy_J,strMetrics,trans_Thermotogae)
model <- lm(trans_cumulative_energy_J ~ I(trans_PI^2)+I(trans_Thermotogae^2)+(.)^2, data = df) %>% step(trace = 0,.)
summary(model)
par(mfrow = c(2, 2))
plot(model)

#After looking at all the squared terms of our predictor variables, it actually makes our model worse!
```

```{r fig.height=9, fig.width=9}
#best (original) model
df <- rsm_hist %>% select(trans_cumulative_energy_J,strMetrics,trans_Thermotogae)
model <- lm(trans_cumulative_energy_J ~ (.)^2, data = df) %>% step(trace = 0,.)
par(mfrow = c(2, 2))
plot(model)
```

#Summed RA for the potential hydrogenotrophs? RSM? Correlates with thermotogae?
```{r}
df <- biom$RA.Otus %>% rownames_to_column(var = "Label") %>% gather(key,value,-Label) %>% filter(key %in% pot_hydrogenotrophs) %>% group_by(Label) %>% summarise(sum_RA = sum(value)) %>% mutate_at(vars(sum_RA),funs(trans_sum_RA_pot_hydr = mutate_transform(.))) %>% select(Label,trans_sum_RA_pot_hydr)
df

#correlates with thermotogae? Negatively
df %>% left_join(rsm_hist,by="Label") %>% ggplot(aes(trans_Thermotogae,trans_sum_RA_pot_hydr))+geom_point()+geom_smooth(method="lm")+stat_cor()


#RSM - doesn't fit significantly to response surface of temp or acetate
rsm_object <- rsm_and_visualize(data = df %>% left_join(rsm_hist,by="Label"),response_cols = "trans_sum_RA_pot_hydr")
rsm_object

#what about gene presence fit to RSM and thermotogae?
df <- ko_metagenome_contributions %>% filter(Gene %in% targetHydrogenaseKO) %>% group_by(OTU) %>% distinct(Gene,.keep_all = TRUE) %>% summarise(hydrGeneCountPerGenome = sum(GeneCountPerGenome)) %>% arrange(desc(hydrGeneCountPerGenome))
# gene presence doesn't correlate significantly with thermotogae
biom$biom_tab %>% rownames_to_column(var = "OTU") %>% select(-match(taxaNames,names(.))) %>% gather(Label,genomes_per_OTU,-OTU) %>% mutate_at(vars(OTU,genomes_per_OTU),funs(as.numeric)) %>% filter(OTU %in% pot_hydrogenotrophs) %>% left_join(.,df,by="OTU") %>% mutate(totalHydrGenes = genomes_per_OTU*hydrGeneCountPerGenome) %>% as_tibble %>% group_by(Label) %>% summarise(sumtotalHydrGenes = sum(totalHydrGenes)) %>% mutate_at(vars(sumtotalHydrGenes),funs(transsumtotalHydrGenes = mutate_transform(.))) %>% 
  left_join(rsm_hist,by="Label") %>% ggplot(aes(trans_Thermotogae,transsumtotalHydrGenes))+geom_point()+geom_smooth(method="lm")+stat_cor()

# hydrogenotrophic gene presence fit to rsm - not significant
biom$biom_tab %>% rownames_to_column(var = "OTU") %>% select(-match(taxaNames,names(.))) %>% gather(Label,genomes_per_OTU,-OTU) %>% mutate_at(vars(OTU,genomes_per_OTU),funs(as.numeric)) %>% filter(OTU %in% pot_hydrogenotrophs) %>% left_join(.,df,by="OTU") %>% mutate(totalHydrGenes = genomes_per_OTU*hydrGeneCountPerGenome) %>% as_tibble %>% group_by(Label) %>% summarise(sumtotalHydrGenes = sum(totalHydrGenes)) %>% mutate_at(vars(sumtotalHydrGenes),funs(transsumtotalHydrGenes = mutate_transform(.))) %>% 
  left_join(rsm_hist,by="Label") %>% rsm_and_visualize(data=.,response_cols = "transsumtotalHydrGenes")
  

```





